go/service/chat_local.go

```diff
@@ -255,55 +422,36 @@ func (h *chatLocalHandler) GetMessagesLocal(ctx context.Context, arg keybase1.Me
 		}
 	}
 
-	var conversations map[chat1.ConversationID]bool
-	if len(arg.Conversations) > 0 {
-		conversations := make(map[chat1.ConversationID]bool)
-		for _, c := range arg.Conversations {
-			conversations[c] = true
-		}
+	if len(arg.Conversations) == 0 {
+		return nil, errors.New("At least 1 conversation ID is required")
 	}
 
 	if arg.Limit <= 0 {
 		arg.Limit = int(^uint(0) >> 1) // maximum int
 	}
 
-	ipagination := &chat1.Pagination{Num: 20}
-getinbox:
-	for i := 0; i < 10000; /* in case we have a server bug */ i++ {
-		iview, err := h.GetInboxLocal(ctx, ipagination)
+	for _, cid := range arg.Conversations {
+		conversationRemote, err := h.remoteClient().GetConversationMetadataRemote(ctx, cid)
 		if err != nil {
-			return nil, err
+			return nil, fmt.Errorf("getting conversation %v error: %v", cid, err)
 		}
-
-		for _, conversation := range iview.Conversations {
-			if conversations != nil && !conversations[conversation.Metadata.ConversationID] {
-				continue
-			}
-			conv, err := h.getConversationMessages(ctx, &conversation, messageTypes, &arg)
-			if err != nil {
-				return nil, err
-			}
-			if len(conv.Messages) != 0 {
-				messages = append(messages, conv)
-
-				if arg.Limit <= 0 {
-					break getinbox
-				}
-			}
+		conversationLocal, err := h.getConversationMessages(ctx, conversationRemote, messageTypes, &arg)
+		if err != nil {
+			return nil, fmt.Errorf("getting messages for conversation %v error: %v", cid, err)
 		}
-
-		// TODO: determine whether need to continue according to the MessageSelector
-		if iview.Pagination == nil || iview.Pagination.Last {
-			break getinbox
-		} else {
-			ipagination = iview.Pagination
+		if len(conversationLocal.Messages) != 0 {
+			conversations = append(conversations, conversationLocal)
+			arg.Limit -= len(conversationLocal.Messages)
+			if arg.Limit <= 0 {
+				break
+			}
 		}
 	}
 
-	return messages, nil
+	return conversations, nil
 }
 
-func (h *chatLocalHandler) fillSenderIDsForPostLocal(arg *keybase1.PostLocalArg) error {
+func (h *chatLocalHandler) fillSenderIDsForPostLocal(arg *keybase1.MessagePlaintext) error {
```

Comment:
The errors below are what I was referring to.

    libkb.LoginRequiredError, libkb.DeviceRequiredError

should be used.

However, this is not how to check if the user is logged in and has a device.  If possible, it would be nice to use an engine with Prereqs{Device: true} to reuse the existing code.  

But if it has to be in the service, then:

```
func IsProvisioned(e Engine, ctx *Context) (bool, error) {
	if ctx.LoginContext != nil {
		return ctx.LoginContext.LoggedInProvisionedLoad()
	}
	return e.G().LoginState().LoggedInProvisionedLoad()
}
```